### ISSUE TITLE
[FR] - Provide comprehensive CDDL grammar for local tx-submission protocol

### ISSUE THREAD

**Issue by ashisherc** [IntersectMBO/cardano-node#5139]
**External**

**Area**

*Other*   Local tx submit mini protocol.

**Describe the feature you'd like**
We are using Local Tx submit mini protocol to submit transactions to the Cardano node, the mini protocol spec does not define the response CDDL as it is understood to be polymorphic. We want to parse the rejection message returned from the Cardano node, for which a CDDL spec would suffice.

**Comment by ashisherc** [comment]
@JaredCorduan

**Comment by JaredCorduan** [comment]
It might make sense to transfer this to https://github.com/input-output-hk/ouroboros-network. what do you think, @coot ?

**Comment by ashisherc** [comment]
here's one for the Ouroboros Network repo, https://github.com/input-output-hk/ouroboros-network/issues/4521

**Comment by coot** [comment]
The problem is that it's indeed polymorphic, and the concrete type is implemented by the consensus team if I am not mistaken.  But they don't use `CDDL`. @dnadales should be able to point to the right place where it's defined.  This is a chase since the rejection reason type itself is polymorphic over blk type (it's a type family: [`ApplyTxErr`](https://github.com/input-output-hk/ouroboros-network/blob/master/ouroboros-consensus/src/Ouroboros/Consensus/Ledger/SupportsMempool.hs?plain=1#L33)), and thus might depend on the era.

**Comment by ashisherc** [comment]
@JaredCorduan @coot in that case, should this issue be moved to the ledger repo? Also, is there a way I can see the specs somehow currently so that I can decode the rejection message (very imp for me at this time)? I also want to emphasise on the importance of documenting this because many teams are now building their own tools in different languages (which is what Cardano needs the most) and we need a way to know the specs.

PS, the Ouroboros network library I am building is in javascript which has gone through multiple iterations and I want to open source it soon.

**Comment by coot** [comment]
> @JaredCorduan @coot in that case, should this issue be moved to the ledger repo? 

Ledger has nothing to do with it, it's consensus (which now lives in `ouroboros-network` repo as well).  I am aware this is important, let's give consensus a chance to answer.

> PS, the Ouroboros network library I am building is in javascript which has gone through multiple iterations and I want to open source it soon.

Nice, I am happy to hear that.

**Comment by abailly-iohk** [comment]
I am not sure this is really consensus specific. Seems to me that what this issue is about is somewhat broader in scope than a mere documentation if we want to do it properly, but it would rather be specifying and exposing the mini-protocols as the node's interface in a more formal way. This would imply:
1. Write, publish, and version the CDDLs somewhere, eg. as a package that implementors could use tool with
2. Write the proper tests to ensure our implementation is compliant with its specification and backward compatible (golden tests, roundtrip properties, version migration tests...)
3. (optional) Publish test vectors or generators that implementors could use to check their implementation against real data

**Comment by abailly-iohk** [comment]
It would be great if we could consolidate the CDDLs we have in, say, a `cardano-cddl-api` package

**Comment by ashisherc** [comment]
@abailly-iohk thanks for the response. I am not sure about another Haskell package for this purpose. (Maybe I misunderstood your message) The sole purpose of my request is so that devs can build tools easily in different languages which just requires us to know the CDDLs.

Just like we have CDDLs documented for Cardano-ledger and ouroboros-network, we would like to have access to all sorts of  CDDLs eg. We also do not have CDDLs for local state query protocol.

I can understand this might be a bigger fish to fry, so at this time I am very much looking for CDDLs for the tx submit rejection message and can wait for other CDDLs

**Comment by abailly-iohk** [comment]
@ashisherc I meant package in the broader sense and I fully sympathise with the goal you describe. In my mind, `package` was intended to mean "something that's published in a well-known location that people can download and use within their environment". It could be as simple as a tarball or just a schema document. the important point is that this "package" would be versioned and could be used within some toolchain to produce, eg. language-dependent artifacts able to serialise/deserialise packets

**Comment by abailly-iohk** [comment]
The CDDL for the rejection messages simply do not exist at this stage so we would need to write them and put them somewhere  for you to retrieve. It's therefore does not seem illogical for me to walk the extra mile and integrate the various CDDLs we have into something more comprehensive. If we were in 2003 we would probably write a XML schema :)

**Comment by ashisherc** [comment]
Perfect!

**Comment by abailly-iohk** [comment]
FWIW the reject reason is an `ApplyTxError` whose encoding is defined here: https://github.com/input-output-hk/cardano-ledger/blob/master/eras/shelley/impl/src/Cardano/Ledger/Shelley/API/Mempool.hs#L214 
Seems to me it would make to include a CDDL file describing those in the ledger itself, alongside the `transaction.cddl` which are defined per era.

**Comment by abailly-iohk** [comment]
@coot @JaredCorduan 

> The problem is that it's indeed polymorphic, and the concrete type is implemented by the consensus team if I am not mistaken.

The concrete type for ledger errors is defined _mostly_ in the ledger, with the consensus only adding a tag identifying the era AFAICT.

**Comment by JaredCorduan** [comment]
> The concrete type for ledger errors is defined _mostly_ in the ledger, with the consensus only adding a tag identifying the era AFAICT.

It is definitely true that the predicate failures in the ledger rules are serialized by the ledger packages, and that this documentation (ie a CDDL for them) does not exist. The same is true for the ledger state serialization.

These failures are wrapped by consensus, and then supplied to the mini-protocols, as best I understand. I believe there are other errors as well, but I'll take @abailly-iohk 's word for it that the majority of them are these predicate failures. I've made an issue for this is the ledger repo. (but note that this is just one of the parts of the overall goal.)

**Comment by abailly-iohk** [comment]
@JaredCorduan Thanks. I would be happy to help building some kind of comprehensive CDDL specification single point of entry that the community could use, but I guess ledger team is best placed to provide consistent CDDL specs of exposed data structures.

**Comment by abailly-iohk** [comment]
There would be some CDDL which would have to be defined in consensus as some error cases are internal there (eg. era mismatch). Perhaps a first step on consensus side would be to write serialisation code checks wrt. CDDL spec mini-protocols payload.

**Comment by JaredCorduan** [comment]
true, the "compressed" part of CBOR is sufficiently obvious to make CBOR "JSON plus tags" (loosely speaking).

**Comment by JaredCorduan** [comment]
I do think that a CDDL spec is the right way to go. but in the mean time, I can drop a  bit of guidance here for anyone in a hurry to understand the predicate failure serialization.

---

The failures are nested in exactly the same structure of the ledger rules, you you can refer to this like [this picture](https://github.com/input-output-hk/cardano-ledger/blob/master/eras/shelley/formal-spec/rules.pdf) to get the tree structure.

Each failure has the following structure:
* a definite list encoding.
* the first field indicates which particular failure for a given ledger rule has fired.
* the remaining fields are specific extra information for that particular failure.

For example, look at the encoding for the failure that results from insufficient fee:

https://github.com/input-output-hk/cardano-ledger/blob/cffa75fdbd800cda60997791e51bf02f2af0c42b/eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Utxo.hs#L216-L220

Even if you do not know haskell, hopefully it is not too hard to see that this is a list of length 3, the identifier within this collection of UTXO failures is `4`, and it also includes two other fields. The other fields are described here:

https://github.com/input-output-hk/cardano-ledger/blob/cffa75fdbd800cda60997791e51bf02f2af0c42b/eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Utxo.hs#L146-L148

I am definitely not asking everyone to learn haskell to get this information, I'm just trying to be helpful right now, knowing it might be a while before we can document everything.

Each predicate failures tends to show up very few times in the codebase, so grep/ag works quite well for finding things. Moreovere, the constructor names, such as `FeeTooSmallUTxO` are exactly what show up in the node logs.

**Comment by abailly-iohk** [comment]
@ashisherc How about reframing this issue as something like:

> Provide comprehensive CDDL grammar for local tx-submission protocol

**Comment by ashisherc** [comment]
!

**Comment by ashisherc** [comment]
!

**Comment by ashisherc** [comment]
!

**Comment by ashisherc** [comment]
!

**Comment by ashisherc** [comment]
!

**Comment by ashisherc** [comment]
!

**Comment by ashisherc** [comment]
!

**Comment by ashisherc** [comment]
!

**Comment by ashisherc** [comment]
!

**Comment by SebastienGllmt** [comment]
!

**Comment by ashisherc** [comment]
!
